Support float#

goldfirere opened the pull request.

goldfirere committed:
Update parser-standard for float#

Signed-off-by: Richard Eisenberg <reisenberg@janestreet.com>

goldfirere committed:
Add support for float#

Signed-off-by: Richard Eisenberg <reisenberg@janestreet.com>

goldfirere committed:
Add tests

But this doesn't work. I'm very confused about how to test here.

Signed-off-by: Richard Eisenberg <reisenberg@janestreet.com>

ccasin commented on lib/Ast.ml:
```diff
@@ -1676,7 +1682,12 @@ end = struct
+          None
+
+      (* Jane Street extension *)
+      | Ptyp_constr_unboxed _ -> Some (Apply, Non)
```
Should we have a `Ptyp_constr_unboxed` case like this `Pstr_constr` case:

```ocaml
 Ptyp_constr (_, _ :: _ :: _) -> Some (Comma, Non)
```

To be forward compatible with parameterized unboxed type constrs?

goldfirere committed:
Some more type tests

Signed-off-by: Richard Eisenberg <reisenberg@janestreet.com>

goldfirere committed:
Test float#

Signed-off-by: Richard Eisenberg <reisenberg@janestreet.com>

ccasin commented on vendor/parser-standard/language_extension.ml:
```diff
@@ -0,0 +1,372 @@
+[@@@warning "-32"] (* ocamlformat removes extension-checking, so there
+                      might be unused values here *)
```
I'm confused at a few things here, but I suspect there are good reasons I'm missing:
- Why isn't this file just a copy of the same file from the compiler?
- I suppose this warning change is needed because some things were removed and some things they depend on remain.  There are good reasons to remove things (e.g., because they aren't needed and depend on things you don't want to import).  But then why not just remove the things triggering warning 32 as well?

goldfirere replied:
This was a misdesign on my part. Instead, I've now vendored in an unedited language_extension module and just called `enable_maximal`. Simpler.

ccasin commented on vendor/parser-extended/lexer.mll:
```diff
@@ -438,24 +585,75 @@ rule token = parse
-      { FLOAT (lit, Some modif) }
-  | (float_literal | hex_float_literal | int_literal) identchar+ as invalid
+
+  (* Jane Street modification *)
```
Are "Jane Street extension" (used elsewhere) and "Jane Street modification" intended to mean different things?  If not, probably pick one.

(I also think it would be fine to just delete these comments, which I'm not sure add much and which might even be confusing if we don't add them for all our previous changes.  But either way is fine, and you seem to like them.)

goldfirere replied:
Yes. the former is a pure addition. The latter is a modification -- some lines changed, not just lines added. It's possible these comments are not serving a purpose, but I'd like to keep them around until the next upstream merge to see if they're useful.

ccasin reviewed (commented):
Looks good - a few questions.

goldfirere committed:
Forward-proof printing of float#

Signed-off-by: Richard Eisenberg <reisenberg@janestreet.com>

goldfirere committed:
Update parser-standard/language_extension

This also removes the differences between ocamlformat's
language_extension and janestreet's. Instead, we now just
call enable_maximal. Seems simpler.

Signed-off-by: Richard Eisenberg <reisenberg@janestreet.com>

ccasin reviewed (approved):
This looks good once you make the CI happy.  Thanks!
